src/main/java/com/google/collegeplanner/servlets/PlannerServlet.java

```diff
@@ -29,35 +35,209 @@
+      nextCourses; // Courses that come after the given course in the graph
+  private HashMap<String, HashSet<String>> corequisites;
+  private HashMap<String, Integer> depths; // depth of a course in the graph
+  private HashMap<String, Integer> credits; // Number of credits for each course
+  private ArrayList<String> courseList;
+  private int totalCredits;
+
   /**
-   * Organizes courses into the given number of semesters
+   * Organizes courses from POST request into a given number of semesters
    */
   @Override
   public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
-    // TODO(ramyabuva): Implement algorithm to split into the number of semesters
     JSONObject body;
+    int numSemesters;
     try {
       body = getBody(request);
-    } catch (ParseException | NullPointerException e) {
+      numSemesters = Integer.parseInt((String) body.get("semesters"));
+    } catch (ParseException | NullPointerException | NumberFormatException e) {
       respondWithError(
           "Invalid body for POST request.", HttpServletResponse.SC_BAD_REQUEST, response);
       return;
     }
-    String strClasses = (String) body.get("selectedClasses");
-    // Returns the courses in a single semester
-    String[][] courses = {strClasses.split(",")};
+    JSONArray selectedClasses = (JSONArray) body.get("selectedClasses");
+
+    constructGraphs(selectedClasses);
+
+    ArrayList<ArrayList<String>> semesters = new ArrayList<>();
+    for (int i = numSemesters; i > 0; i--) {
+      int avgCredits = totalCredits / i;
+      ArrayList<String> semester = new ArrayList<>();
+      for (int j = 0; j < courseList.size(); j++) {
+        String course = courseList.get(j);
+        if (indegree.get(course) == 0) {
+          // Check if all corequisites also have an indegree of 0
+          if (isValidToAdd(course)) {
+            // If valid, add all to semester
+            courseList.remove(j);
+            j--;
+            avgCredits -= credits.get(course);
+            totalCredits -= credits.get(course);
+            semester.add(course);
+            for (String coreq : corequisites.get(course)) {
+              courseList.remove(coreq);
+              avgCredits -= credits.get(coreq);
+              totalCredits -= credits.get(coreq);
+              semester.add(coreq);
+            }
+          }
+        }
+        // Breaks out if there are no more required classes and we exceeded our credit limit
+        if (depths.get(course) != i && avgCredits <= 0) {
+          break;
+        }
+      }
+      // Update indegrees for all next courses
+      for (String course : semester) {
+        for (String nextCourse : nextCourses.get(course)) {
+          indegree.replace(nextCourse, indegree.get(nextCourse) - 1);
+        }
+      }
+      semesters.add(semester);
+    }
+    // If courseList is not empty, scheduling was not possible
+    if (!courseList.isEmpty()) {
```

Comment:
minor point, but i think it's ok to return status code 200 with 0 results in this case.